### PR TITLE
Add support for multiple preshared keys

### PR DESCRIPTION
**Pull request by josephschorr**
Fixes https://github.com/authzed/spicedb/issues/288
Fixes AUTHZ-500

### CODE REVIEW

File: pkg/cmd/server/server.go [hunk 1]
```diff
@@ -339,7 +352,7 @@ type completedServerConfig struct {
 
 	unaryMiddleware     []grpc.UnaryServerInterceptor
 	streamingMiddleware []grpc.StreamServerInterceptor
-	presharedKey        string
+	presharedKey        []string
```
Comment (jzelinskie) [inline pkg/cmd/server/server.go:355]: presharedKey => presharedKeys

File: pkg/cmd/server/server.go [hunk 2]
```diff
@@ -103,7 +103,11 @@ func (c *Config) Complete() (RunnableServer, error) {
 	}
 
 	if c.GRPCAuthFunc == nil {
-		log.Trace().Int("preshared-key-length", len(c.PresharedKey)).Msg("using gRPC auth with preshared key")
+		log.Trace().Int("preshared-keys", len(c.PresharedKey)).Msg("using gRPC auth with preshared key(s)")
```
Comment (jzelinskie) [inline pkg/cmd/server/server.go:106]: preshared-keys-count

File: pkg/cmd/server/server.go [hunk 3]
```diff
@@ -141,10 +145,15 @@ func (c *Config) Complete() (RunnableServer, error) {
 			return nil, fmt.Errorf("failed to create dispatcher: %w", cerr)
 		}
 
+		presharedKey := ""
```
Comment (jzelinskie) [inline pkg/cmd/server/server.go:148]: presharedKey  => dispatchKey

File: pkg/cmd/server/server.go [hunk 4]
```diff
@@ -141,10 +145,15 @@ func (c *Config) Complete() (RunnableServer, error) {
...
+		if len(c.PresharedKey) > 0 {
```
Comment (jzelinskie) [inline pkg/cmd/server/server.go:149]: Why check this? Shouldn't it error before here if `len(c.PresharedKeys) == 0`?

File: pkg/cmd/server/server.go [hunk 5]
```diff
@@ -103,7 +103,11 @@ func (c *Config) Complete() (RunnableServer, error) {
 	}
```
Comment (jzelinskie) [inline pkg/cmd/server/server.go:103]: here we should check that each preshared key is not the empty string

File: cmd/spicedb/serve_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+				)
+			}
+			require.NoError(err)
+			defer conn.Close()
+
+			resp, err := healthpb.NewHealthClient(conn).Check(context.Background(), &healthpb.HealthCheckRequest{Service: "authzed.api.v1alpha1.SchemaService"})
+			require.NoError(err)
+			require.Equal(healthpb.HealthCheckResponse_SERVING, resp.GetStatus())
+
+			client := v1alpha1.NewSchemaServiceClient(conn)
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:63]: why not use v1?

File: cmd/spicedb/serve_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+			var conn *grpc.ClientConn
+			var err error
+			if key == "" {
+				conn, err = grpc.Dial(fmt.Sprintf("localhost:%s", tester.port), grpc.WithTransportCredentials(insecure.NewCredentials()))
+			} else {
+				conn, err = grpc.Dial(fmt.Sprintf("localhost:%s", tester.port),
+					grpc.WithInsecure(),
+					grpcutil.WithInsecureBearerToken(key),
+				)
+			}
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:55]:
```go
opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
if key != "" {
    opts = append(opts, grpcutil.WithInsecureBearerToken(key))
}
conn, err := grpc.Dial(fmt.Sprintf("localhost:%s", tester.port), opts...)
```

**Review by jzelinskie — CHANGES_REQUESTED** [review]

File: pkg/cmd/server/server.go (hunk 4 above)
Comment (josephschorr) [reply pkg/cmd/server/server.go:149]: Not if no keys were given; remember: preshared keys are ignored if a custom auth function is specified

File: cmd/spicedb/serve_integration_test.go (hunk 6 above)
Comment (josephschorr) [reply cmd/spicedb/serve_integration_test.go:63]: Copy paste

File: .github/workflows/build-test.yaml [hunk 8]
```diff
@@ -50,7 +50,10 @@ jobs:
       - uses: "actions/setup-go@v3"
         with:
           go-version: "^1.17"
-      - uses: "authzed/actions/docker-build@main"
+      - uses: "docker/build-push-action@v1"
```
Comment (jzelinskie) [inline .github/workflows/build-test.yaml:53]: probably need to rename this job

`image-build` => `docker-tests`
`Build Container Image` => `Build & Test Container Image`

File: cmd/spicedb/serve_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	"google.golang.org/grpc/status"
+)
+
+func TestServe(t *testing.T) {
+	requireParent := require.New(t)
+
+	tester, err := newTester(t,
+		&dockertest.RunOptions{
+			Repository:   "authzed/spicedb",
+			Tag:          "currentpr",
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:28]: dev

File: cmd/spicedb/serve_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	defer tester.cleanup()
+
+	for key, expectedWorks := range map[string]bool{
+		"":           false,
+		"firstkey":   true,
+		"secondkey":  true,
+		"anotherkey": false,
+	} {
+		t.Run(key, func(t *testing.T) {
+			require := require.New(t)
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:44]: Not required, but fwiw I prefer to just use require directly and not shadow the import.

File: cmd/spicedb/serve_integration_test.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+			opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+			if key != "" {
+				opts = append(opts, grpcutil.WithInsecureBearerToken(key))
+			}
+			conn, err := grpc.Dial(fmt.Sprintf("localhost:%s", tester.port), opts...)
+
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:51]: why the new line here?

File: cmd/spicedb/serve_integration_test.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+			require.NoError(err)
+			defer conn.Close()
+
+			resp, err := healthpb.NewHealthClient(conn).Check(context.Background(), &healthpb.HealthCheckRequest{Service: "authzed.api.v1alpha1.SchemaService"})
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:55]: This is against v1alpha1, but then you use v1

File: cmd/spicedb/serve_integration_test.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+				}
+				`,
+			})
+
+			if expectedWorks {
+				require.NoError(err)
+			} else {
+				s, ok := status.FromError(err)
+				require.True(ok)
+				require.Equal(codes.Unknown, s.Code())
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:78]: Shouldn't this be `codes.Unauthenticated`?

File: cmd/spicedb/servetesting_integration_test.go [hunk 14]
```diff
@@ -29,7 +29,7 @@ func TestTestServer(t *testing.T) {
 	tester, err := newTester(t,
 		&dockertest.RunOptions{
 			Repository:   "authzed/spicedb",
-			Tag:          "latest",
+			Tag:          "currentpr",
```
Comment (jzelinskie) [inline cmd/spicedb/servetesting_integration_test.go:32]: dev

File: cmd/spicedb/serve_integration_test.go [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+			defer conn.Close()
+
+			resp, err := healthpb.NewHealthClient(conn).Check(context.Background(), &healthpb.HealthCheckRequest{Service: "authzed.api.v1alpha1.SchemaService"})
+			require.NoError(err)
+			require.Equal(healthpb.HealthCheckResponse_SERVING, resp.GetStatus())
+
+			client := v1.NewSchemaServiceClient(conn)
+			_, err = client.WriteSchema(context.Background(), &v1.WriteSchemaRequest{
+				Schema: `
+				definition user {}
```
Comment (jzelinskie) [inline cmd/spicedb/serve_integration_test.go:62]: Why not use a simpler schema? It's not really relevant to the test

File: pkg/cmd/server/server.go (hunk 4 above)
Comment (jzelinskie) [reply pkg/cmd/server/server.go:153]: Oh, gotcha.

I wonder if it makes more sense to hoist this up to where all of the other preshared key logic is.

**Review by jzelinskie — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test.yaml [hunk 16]
```diff
@@ -50,7 +50,10 @@ jobs:
...
+        with:
+          push: false
+          tags: "currentpr"
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply .github/workflows/build-test.yaml:56]: `dev` has a risk of overloading with a, well, development image. I wanted to be explicitly different from all other contexts

File: .github/workflows/build-test.yaml (hunk 8 above)
Comment (josephschorr) [reply .github/workflows/build-test.yaml:53]: If we do, it'll break the CI submit checks, so I left it as is (unless we should change those too)

File: cmd/spicedb/serve_integration_test.go (hunk 10 above)
Comment (josephschorr) [reply cmd/spicedb/serve_integration_test.go:44]: I do too, but I was keeping to the style in the existing integration tests

File: cmd/spicedb/serve_integration_test.go (hunk 13 above)
Comment (josephschorr) [reply cmd/spicedb/serve_integration_test.go:78]: Currently, we don't return that code if the key does not match

File: cmd/spicedb/serve_integration_test.go (hunk 12 above)
Comment (josephschorr) [reply cmd/spicedb/serve_integration_test.go:55]: Its mostly to make sure the service is running at all, but I can change it

File: cmd/spicedb/serve_integration_test.go (hunk 15 above)
Comment (josephschorr) [reply cmd/spicedb/serve_integration_test.go:62]: Copy-pasted it; I can make it simpler, but it wasn't a big deal to change either way

File: .github/workflows/build-test.yaml (hunk 8 above)
Comment (jzelinskie) [reply .github/workflows/build-test.yaml:53]: It's really easy to change them. I'd rather spend two seconds in the GitHub settings than live with names and descriptions of things that don't make sense.

File: cmd/spicedb/serve_integration_test.go (hunk 13 above)
Comment (jzelinskie) [reply cmd/spicedb/serve_integration_test.go:78]: that seems... like a bug

**Review by jzelinskie — APPROVED** [review]

**Pull request merged**